src/core/batch.rkt

```diff
@@ -145,6 +153,86 @@
       [_ node]))
   (unmunge reg))
 
+(define (batch-restore-index batch)
+  (make-hash (for/list ([node (in-vector (batch-nodes batch))]
+                        [n (in-naturals)])
+               (cons node n))))
+
+(define (egg-nodes->batch egg-nodes id->spec input-batch rename-dict)
+  (define out (make-mutable-batch))
+  (set-mutable-batch-index! out (batch-restore-index input-batch))
+
+  ; This fuction here is only because of cycles in loads:( Can not be imported from egg-herbie.rkt
+  (define (egg-parsed->expr expr rename-dict type)
+    (let loop ([expr expr]
+               [type type])
+      (match expr
+        [(? number?) (if (representation? type) (literal expr (representation-name type)) expr)]
+        [(? symbol?)
+         (if (hash-has-key? rename-dict expr) (car (hash-ref rename-dict expr)) (list expr))]
+        [(list '$approx spec impl)
+         (define spec-type (if (representation? type) (representation-type type) type))
+         (approx (loop spec spec-type) (loop impl type))]
+        [(list 'if cond ift iff)
+         (if (representation? type)
+             (list 'if (loop cond (get-representation 'bool)) (loop ift type) (loop iff type))
+             (list 'if (loop cond 'bool) (loop ift type) (loop iff type)))]
+        [(list (? impl-exists? impl) args ...) (cons impl (map loop args (impl-info impl 'itype)))]
+        [(list op args ...) (cons op (map loop args (operator-info op 'itype)))])))
+
+  (define (eggref id)
+    (cdr (vector-ref egg-nodes id)))
+
+  (define (add-root node type)
```

Comment:
I think this is basically `egg-parsed->expr` except without the recursion and with `batch-push!` inserted in? We can rewrite this a nicer way.

Reply:
Yes, it is, `add-root` also needs to access nodes by indexes from so-called `egg-nodes`, just minor detail.
But overall yes, they are the same.
I am keeping `egg-parsed->expr` around to parse `spec`, since spec is an `Expr`, otherwise `add-root` does all the job

Reply:
Cool. I think we can maybe merge them later on but it's not a big deal right now.